replication-api/src/main/java/com/connexta/replication/queue/Task.java

```diff
@@ -0,0 +1,154 @@
+   * returns <code>0</code>).
+   *
+   * @return the time when this task was queued in milliseconds from the epoch
+   */
+  public long getQueuedTime();
+
+  /**
+   * Gets the total number of times this particular task has been attempted. When a task fails for a
+   * reason that doesn't preclude it from being re-tried and it gets re-queued by the worker, its
+   * total attempts counter will automatically be incremented as it is re-added to the associated
+   * queue for later reprocessing.
+   *
+   * @return the total number of times this task has been attempted so far
+   */
+  public int getTotalAttempts();
+
+  /**
+   * Checks if this task is still locked by the current thread/worker. A task will automatically be
+   * unlocked once one of {@link #unlock()}, {@link #complete()}, {@link #fail()}, {@link
+   * #offerForRetry(PriorityLevel)}, {@link #offerForRetry(PriorityLevel, long, TimeUnit)}, or
+   * {@link #putForRetry(PriorityLevel)} has been called on it.
+   *
+   * @return <code>true</code> if the current thread/worker owns the lock to this task; <code>false
+   *     </code> otherwise
+   */
+  public boolean isLocked();
+
+  /**
+   * Attempts to unlock this task. An unlock task will be picked up by the next worker attempting to
+   * poll for a task from the associated queue.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not current locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void unlock() throws QueueException, InterruptedException;
+
+  /**
+   * Reports the successful completion of this task. The task will automatically be unlocked and
+   * removed from the associated queue.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not currently or no longer locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void complete() throws QueueException, InterruptedException;
+
+  /**
+   * Reports the failed completion of this task. The task will automatically be unlocked and removed
+   * from the associated queue. Such a task will not be re-queued for another attempt at being
+   * processed.
+   *
+   * @throws IllegalStateException if the current thread/worker doesn't own the lock to the task or
+   *     if the task is not currently or no longer locked
+   * @throws InterruptedException if the thread was interrupted while attempting to unlock the task
+   * @throws QueueException if an error occurred while performing the operation
+   */
+  public void fail() throws QueueException, InterruptedException;
```

Comment:
❓ Would it be worth passing a reason string for better reporting experience? May not be needed right away.

Reply:
These were changed after an offline discussions.